process/sync/metablock.go

```diff
@@ -157,46 +155,165 @@ func (boot *MetaBootstrap) removeBlockBody(
 	}
 
-	header, err := process.GetShardHeaderFromStorage(headerHash, boot.marshalizer, boot.store)
-	if err != nil {
-		return err
+	metaBlock, ok := headerHandler.(*block.MetaBlock)
+	if !ok {
+		log.Info(process.ErrWrongTypeAssertion.Error())
+		return nil, false
+	}
+
+	if metaBlock.Round > boot.bootstrapRoundIndex {
+		log.Info(ErrHigherRoundInBlock.Error())
+		return nil, false
 	}
 
-	log.Info(fmt.Sprintf("apply notarized block with nonce %d and round %d\n", header.Nonce, header.Round))
+	return metaBlock, true
+}
+
+func (boot *MetaBootstrap) getMaxNotarizedHeadersNoncesInMetaBlock(
+	metaBlock *block.MetaBlock,
+	ni *notarizedInfo,
+) (map[uint32]uint64, error) {
 
-	if header.GetRound() > boot.boostrapRoundIndex {
-		return ErrHigherBootstrapRound
+	maxNonce := make(map[uint32]uint64, 0)
+	for _, shardData := range metaBlock.ShardInfo {
+		header, err := process.GetShardHeaderFromStorage(shardData.HeaderHash, boot.marshalizer, boot.store)
+		if err != nil {
+			return maxNonce, err
+		}
+
+		if header.Nonce > maxNonce[shardData.ShardId] {
+			maxNonce[shardData.ShardId] = header.Nonce
+		}
 	}
 
-	boot.blkExecutor.SetLastNotarizedHdr(header.ShardId, header)
+	return maxNonce, nil
+}
 
-	return nil
+func (boot *MetaBootstrap) areNotarizedShardHeadersFound(
+	ni *notarizedInfo,
+	notarizedNonce map[uint32]uint64,
+	nonce uint64,
+) bool {
+
+	for i := uint32(0); i < boot.shardCoordinator.NumberOfShards(); i++ {
+		if ni.lastNotarized[i] == 0 {
+			ni.lastNotarized[i] = notarizedNonce[i]
+			ni.nonceWithLastNotarized[i] = nonce
+			continue
+		}
+
+		if ni.finalNotarized[i] == 0 {
+			ni.finalNotarized[i] = notarizedNonce[i]
+			ni.nonceWithFinalNotarized[i] = nonce
+			continue
+		}
+	}
+
+	foundAllNotarizedShardHeaders := true
+	for i := uint32(0); i < boot.shardCoordinator.NumberOfShards(); i++ {
+		if ni.lastNotarized[i] == 0 || ni.finalNotarized[i] == 0 {
+			foundAllNotarizedShardHeaders = false
+			break
+		}
+	}
+
+	return foundAllNotarizedShardHeaders
 }
 
-func (boot *MetaBootstrap) getHeaderFromStorage(nonce uint64) (data.HeaderHandler, []byte, error) {
-	nonceToByteSlice := boot.uint64Converter.ToByteSlice(nonce)
-	headerHash, err := boot.store.Get(dataRetriever.MetaHdrNonceHashDataUnit, nonceToByteSlice)
-	if err != nil {
-		return nil, nil, err
+func (boot *MetaBootstrap) getNonceWithLastNotarized(nonce uint64) (uint64, map[uint32]uint64, map[uint32]uint64) {
+	ni := notarizedInfo{}
+	ni.reset()
+	for currentNonce := nonce; currentNonce > 0; currentNonce-- {
+		metaBlock, ok := boot.isMetaBlockValid(currentNonce)
+		if !ok {
+			ni.reset()
+			continue
+		}
+
+		if ni.startNonce == 0 {
+			ni.startNonce = currentNonce
+		}
+
+		if len(metaBlock.ShardInfo) == 0 {
+			continue
+		}
+
+		maxNonce, err := boot.getMaxNotarizedHeadersNoncesInMetaBlock(metaBlock, &ni)
+		if err != nil {
+			log.Info(err.Error())
+			ni.reset()
+			continue
+		}
+
+		if !boot.areNotarizedShardHeadersFound(&ni, maxNonce, currentNonce) {
```

Comment:
suggestion to change if boot.arenotarized.. == true -> break.

Reply:
Done